lib/puppet/file_serving/terminus_helper.rb

```diff
@@ -29,4 +29,16 @@ def path2instances(request, *paths)
       path2instance(request, base_path, :ignore_source_permissions => true, :relative_path => file)
     end
   end
+
+  def self.unescape_url(url)
+    url.sub(/^url=/,'')
+  end
+
+  def self.escape_url(url)
+    if url =~ /^https?:/
+      "url=#{url}"
+    else
+      url
+    end
```

Comment:
Ah so the line https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/request.rb#L253 is completely wrong! We no longer send the environment in the URI path, it's now sent as a query parameter.

I suspect you'd have to do something like https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/request.rb#L248-L251 to short-circuit evaluation for http(s) protocols.

Ideally puppet should not mutate the request key like the way it does. I'd prefer to see the request preserve the original URL, and extract the parts it needs, e.g. 

```
def protocol
  @uri.scheme
end
```

That would be more work than in this PR though.